src/lib/react/useLocalMedia.ts

```diff
@@ -1,25 +1,119 @@
+              speakerDevices: MediaDeviceInfo[];
+          };
+      }
+    | {
+          type: "LOCAL_STREAM_UPDATED";
+          payload: { stream: MediaStream; currentCameraDeviceId?: string; currentMicrophoneDeviceId?: string };
+      };
+
+export type LocalMediaRef = { state: LocalMediaState; actions: LocalMediaActions; _ref: LocalMedia | null };
+
+function reducer(state: LocalMediaState, action: LocalMediaEvents): LocalMediaState {
+    switch (action.type) {
+        case "DEVICE_LIST_UPDATED":
+            return {
+                ...state,
+                ...action.payload,
+            };
+        case "LOCAL_STREAM_UPDATED":
+            return {
+                ...state,
+                currentCameraDeviceId: action.payload.currentCameraDeviceId,
+                currentMicrophoneDeviceId: action.payload.currentMicrophoneDeviceId,
+                localStream: action.payload.stream,
+            };
+        default:
+            return state;
+    }
+}
+
+export default function useLocalMedia(
+    constraints: MediaStreamConstraints = { audio: true, video: true }
+): LocalMediaRef {
+    const [localMedia, setLocalMedia] = useState<LocalMedia | null>(null);
+    const [state, dispatch] = useReducer(reducer, initialState);
 
     useEffect(() => {
-        const getLocalStream = async () => {
-            if (!localStream) {
-                const stream = await navigator.mediaDevices.getUserMedia({ audio: true, video: true });
-                setLocalStream(stream);
-            }
-        };
+        setLocalMedia(new LocalMedia(constraints));
+    }, []);
 
-        getLocalStream();
+    useEffect(() => {
+        if (!localMedia) {
+            return;
+        }
 
-        // Stop tracks on unmount
-        return () => {
-            localStream?.getTracks().forEach((t) => {
-                t.stop();
+        localMedia.addEventListener("device_list_updated", (e) => {
+            const { cameraDevices, microphoneDevices, speakerDevices } = e.detail;
+            dispatch({ type: "DEVICE_LIST_UPDATED", payload: { cameraDevices, microphoneDevices, speakerDevices } });
+        });
+
+        localMedia.addEventListener("stream_updated", (e) => {
+            const { stream } = e.detail;
+            dispatch({
+                type: "LOCAL_STREAM_UPDATED",
+                payload: {
+                    stream,
+                    currentCameraDeviceId: localMedia.getCameraDeviceId(),
+                    currentMicrophoneDeviceId: localMedia.getMicrophoneDeviceId(),
+                },
             });
+        });
+
+        localMedia.start();
+
+        // Perform cleanup on unmount
+        return () => {
+            localMedia?.stop();
         };
-    }, [localStream]);
+    }, [localMedia]);
 
-    return [localStream];
+    return {
+        state,
+        actions: {
+            setCameraDevice: (...args) => {
+                return localMedia?.setCameraDevice(...args) || Promise.resolve();
```

Comment:
If you set the initial value in `useState`, `localMedia` should be available here always, or am I missing something?

Reply:
Yes, it should be. Ill try tweaking this, as it is much cleaner without the optional chaining.

Reply:
Fixed